contracts/trading/seaport/zones/ImmutableSignedZoneV2.sol

```diff
@@ -0,0 +1,572 @@
+            revert InvalidExtraData("invalid context, unexpected context length", zoneParameters.orderHash);
+        }
+    }
+
+    /**
+     * @dev Validates substandard 3.
+     *
+     * @param context Bytes payload of context, 0 indexed to start of substandard segment.
+     * @param zoneParameters The zone parameters.
+     * @return Length of substandard segment.
+     */
+    function _validateSubstandard3(bytes calldata context, ZoneParameters calldata zoneParameters)
+        internal
+        pure
+        returns (uint256)
+    {
+        if (uint8(context[0]) != 3) {
+            return 0;
+        }
+
+        if (context.length < 33) {
+            revert InvalidExtraData("invalid substandard 3 data length", zoneParameters.orderHash);
+        }
+
+        if (_deriveReceivedItemsHash(zoneParameters.consideration, 1, 1) != bytes32(context[1:33])) {
+            revert Substandard3Violation(zoneParameters.orderHash);
+        }
+
+        return 33;
+    }
+
+    /**
+     * @dev Validates substandard 4.
+     *
+     * @param context Bytes payload of context, 0 indexed to start of substandard segment.
+     * @param zoneParameters The zone parameters.
+     * @return Length of substandard segment.
+     */
+    function _validateSubstandard4(bytes calldata context, ZoneParameters calldata zoneParameters)
+        internal
+        pure
+        returns (uint256)
+    {
+        if (uint8(context[0]) != 4) {
+            return 0;
+        }
+
+        // substandard ID + array offset + array length
+        if (context.length < 65) {
+            revert InvalidExtraData("invalid substandard 4 data length", zoneParameters.orderHash);
+        }
+
+        uint256 expectedOrderHashesSize = uint256(bytes32(context[33:65]));
+        uint256 substandardIndexEnd = 64 + (expectedOrderHashesSize * 32);
+        bytes32[] memory expectedOrderHashes = abi.decode(context[1:substandardIndexEnd + 1], (bytes32[]));
+
+        // revert if any order hashes in substandard data are not present in zoneParameters.orderHashes
+        if (!_bytes32ArrayIncludes(zoneParameters.orderHashes, expectedOrderHashes)) {
+            revert Substandard4Violation(zoneParameters.orderHashes, expectedOrderHashes, zoneParameters.orderHash);
+        }
+
+        return substandardIndexEnd + 1;
+    }
+
+    /**
+     * @dev Validates substandard 6.
+     *
+     * @param context Bytes payload of context, 0 indexed to start of substandard segment.
+     * @param zoneParameters The zone parameters.
+     * @return Length of substandard segment.
+     */
+    function _validateSubstandard6(bytes calldata context, ZoneParameters calldata zoneParameters)
+        internal
+        pure
+        returns (uint256)
+    {
+        if (uint8(context[0]) != 6) {
+            return 0;
+        }
+
+        if (context.length < 65) {
+            revert InvalidExtraData("invalid substandard 6 data length", zoneParameters.orderHash);
+        }
+
```

Comment:
I'd like to see some comments here around how this all works and why

Reply:
https://github.com/immutable/contracts/pull/189/commits/e73ff8731d0de8588e3a64fe0b1230f64d891948